test/externalTests/gnosis.sh

```diff
@@ -47,7 +47,7 @@ function gnosis_safe_test
         "${compile_only_presets[@]}"
         #ir-no-optimize            # Compilation fails with "YulException: Variable var_call_430_mpos is 1 slot(s) too deep inside the stack."
         #ir-optimize-evm-only      # Compilation fails with "YulException: Variable var_call_430_mpos is 1 slot(s) too deep inside the stack."
-        ir-optimize-evm+yul
+        #ir-optimize-evm+yul       # Compilation fails with "YulException: Cannot swap Variable var_operation with Variable _1: too deep in the stack by 4 slots."
```

Comment:
Are we tracking this in the Gnosis repo, i.e. are we hoping this gets fixed upstream at a certain point in time?

Reply:
I believe the fix depends more on Hardhat adding proper support to viaIR. Although, as far as I understood, Gnosis could annotate `memory-[un]safe` where needed in their assembly code blocks to avoid such error with this compiler preset. I'm not sure, but maybe Gnosis is not targeting yul optimizer by default? Otherwise, I think they would already had gotten such error as well.

Reply:
I agree, we should at least submit an issue about it upstream. Or, better, a PR fixing this as we did in the past. I think almost no one is targeting via IR at this point, but they do sometimes accept PRs from us fixing the problems. Having to disable this preset greatly reduces the usefulness of this external test for us.

Reply:
By not using `"memory-safe"` blocks in cases where the assembly is actually memory safe they are potentially missing out on some optimizations.

Reply:
> maybe Gnosis is not targeting yul optimizer by default?

At Safe we're targeting solc 0.7.6 with optimizer disable

Reply:
> I believe the fix depends more on Hardhat adding proper support to viaIR.

Why would it depend on hardhat? I get the same error using the solc cli

Reply:
Yeah, I don't think it depends on anything in Hardhat. @r0qs is probably thinking of https://github.com/NomicFoundation/hardhat/issues/3365 - there were quite a few things that broke on Hardhat's side because its heuristics were assuming things that were only true with the legacy codegen.

What will solve this properly is probably #13972, which will enable stack optimization even when the user requests optimizer to be disabled. This will give you unoptimized Yul output, with just the stack optimization applied when Yul is converted to EVM bytecode.

Still, in the short term this failure might be solved simply by using to `"memory-safe"` assembly blocks where appropriate (that was @ekpyron's suggestion), which should be done anyway.

Reply:
> Still, in the short term this failure might be solved simply by using to "memory-safe" assembly blocks where appropriate (that was @ekpyron's suggestion), which should be done anyway.

Do you mean that this should be adjusted in the Safe source code? The [function](https://github.com/safe-global/safe-contracts/blob/cb4b2b19b3e336b8defd3b8c9e0e6a2ae130598c/contracts/Safe.sol#L139-L150) where the error allegedly happens doesn't make use of assembly

Reply:
So it seems that adding `indexed` to event parameters broke the thing: https://github.com/safe-global/safe-contracts/commit/b9fdbde766eb0fee958cb9b12a3b3e5193fd96ae#diff-587b494ea631bb6b7adf4fc3e1a2e6a277a385ff16e1163b26e39de24e9483deR62-R63

If I remove the keyword, it works fine again

Reply:
> @r0qs Did you submit the issue/PR? That's the only blocker here.

No I did not. Will do it shortly.

> Yeah, I don't think it depends on anything in Hardhat. @r0qs is probably thinking of [NomicFoundation/hardhat#3365](https://github.com/NomicFoundation/hardhat/issues/3365) - there were quite a few things that broke on Hardhat's side because its heuristics were assuming things that were only true with the legacy codegen.

Exactly, I thought it could be related. 

> Do you mean that this should be adjusted in the Safe source code? The [function](https://github.com/safe-global/safe-contracts/blob/cb4b2b19b3e336b8defd3b8c9e0e6a2ae130598c/contracts/Safe.sol#L139-L150) where the error allegedly happens doesn't make use of assembly

Yes, this is what @cameel meant. Also, this function does seems to use assembly, not directly, but through [checkNSignatures](https://github.com/safe-global/safe-contracts/blob/cb4b2b19b3e336b8defd3b8c9e0e6a2ae130598c/contracts/Safe.sol#L304). But I think that the problem is that the project will not even compile when using `0.8.19` with the yul optimizer enabled, not exclusively because of that specific function, but all others that also make use of assembly, potentially in an "unsafe" way.

You should annotate any inline assembly block that either access memory or assigns to any solidity variables in memory (reference from the docs [here](https://docs.soliditylang.org/en/v0.8.19/assembly.html#memory-safety)), for instance if you make use of `mload` or `mstore`, those could be "memory unsafe" and thus you need to properly annotate them.

> So it seems that adding `indexed` to event parameters broke the thing: [safe-global/safe-contracts@b9fdbde#diff-587b494ea631bb6b7adf4fc3e1a2e6a277a385ff16e1163b26e39de24e9483deR62-R63](https://github.com/safe-global/safe-contracts/commit/b9fdbde766eb0fee958cb9b12a3b3e5193fd96ae#diff-587b494ea631bb6b7adf4fc3e1a2e6a277a385ff16e1163b26e39de24e9483deR62-R63)
> 
> If I remove the keyword, it works fine again

Yes, this is the strange part for me as well, I don't get why the `indexed` would break this to be honest.

Reply:
> Yes, this is the strange part for me as well, I don't get why the `indexed` would break this to be honest.

Making an event parameter `indexed` means that the compiler now adds a topic. So for example for an event with one parameter it makes it switch from `LOG1` which accepts 3 values on the stack to `LOG2`, which accepts 4. This leaves one slot less available for other things on the stack.

Still, the fact that this resolves the problem is a good hint. If you want to keep it `indexed`, you could instead play around the places that `emit` it and try to reduce the number of local variables or parameters. Even just changing the order of things or wrapping some blocks of code into `{}` could resolve it.

But yeah, updating the `assembly` blocks would be even better if that works for you. Note that memory-unsafe assembly results in disabling stack optimizations for the whole contract, not just the single path where it's used.

Reply:
In any case, @r0qs here we just need a TODO comment with a link to issue that will make the gnosis ext test usable to us again and we can merge it. This can either be an upstream issue (if Safe is willing to accept a short-term workaround) or #13972 if we have to wait for the full solution on our side.

As for the workaround, would be best if you could submit a PR that changes the assembly blocks to memory-safe (provided that they actually are memory safe). You can use the [magic `/// @solidity` comment](https://docs.soliditylang.org/en/latest/assembly.html#memory-management) instead of the `("memory-safe")` annotation to have the code still compile on 0.7.6. That PR itself is not a blocker to merging this though.

Reply:
Actually, a small correction. @ekpyron rightly pointed out that this is happening with full optimizer enabled, so #13972 is not really a solution. It's not something that will be fixed on compiler side because the compiler already has a mechanism to deal with it, it's simply that memory-unsafe assembly prevents that mechanism from kicking in. This will require a change on the Safe side.